src/index.ts

```diff
@@ -25,17 +25,16 @@ export class Route {
     const route = url.parse(s, true)
     const parts = fromNullable(route.pathname)
       .map(s =>
-        s
+        decodeURI(s)
           .split('/')
           .filter(Boolean)
-          .map(decodeURIComponent)
       )
       .getOrElse([])
     return new Route(parts, route.query)
   }
   toString(): string {
     const qs = querystring.stringify(this.query)
-    return '/' + this.parts.map(encodeURIComponent).join('/') + (qs ? '?' + qs : '')
+    return encodeURI('/' + this.parts.join('/')) + (qs ? '?' + qs : '')
```

Comment:
`encodeURI` does not encode as many characters as `encodeURIComponent`. For example, it does not encode `?`. This means a route containing `?` will not be encoded properly, and it would be mistaken for the start of a query string.

Example: 

``` js
encodeURI('/search/foo?bar') // => not desired: '/search/foo?bar'
`/search/${encodeURIComponent('foo?bar')}` // => desired: '/search/foo%3Fbar'
``

Reply:
Using a `?` in `parts` seems weird, isn't it? In that case we could require the user to escape manually.
Alternatively what do you propose to get both the desired behaviours (`@` and `?`)?

Reply:
Not necessarily, in this case the path is generated from a search query. The user could type anything into the search input, including other reserved characters such as `@`, `&`, which would not be encoded under `encodeURI` but would be under `encodeURIComponent`.

Ideally path parts would be encoded/decoded as components by default, but with an option to disable or override this.

Slightly related: we also need to do some custom encoding, as we wish to encode spaces as `-` characters instead of `%20` (slugify). I'm currently doing this by slugifying the value before passing it to `format`, and then letting `format` URI encode the rest. E.g.

``` ts
format(search, { query: slugify('foo & bar') }) // => '/search/foo-%26-bar'
```

I'm not sure what the best solution is, but hopefully these requirements make sense.

Reply:
Of course one solution to this is if fp-ts-routing didn't enforce any encoding/decoding and this could all be handled in user land (decoding when parsing, encoding before formatting). If we decided on this we would also have to raise the same question for the query params.

Reply:
> but with an option to disable or override this

The current implementation (with `encodeURI`) seems a good default though.
If you must override the default behaviour using an option then I'm not sure what's the difference from requiring the user to escape manually (or apply a transformation as your slugify) in the few cases where it matters. In both cases, as a user, you must remember to do something, however adding an option makes the API more complicated without providing more safety

Reply:
> escape manually (or apply a transformation as your slugify)

Without a way to disable the in-built encoding, any user defined encoding would lead to double encoding. For example:

``` ts
encodeURI(encodeURIComponent('foo & bar ')) // => "foo%2520%2526%2520bar%2520"
```

Reply:
I see, thanks for explaining

Reply:
@gcanti Given that, where do you think we stand with #18?

Should we just remove the in-built encoding/decoding, and let the user handle it where it's necessary? 🤔